src/3d/symbols/qgspoint3dbillboardmaterial.cpp

```diff
@@ -0,0 +1,171 @@
+  : mSize( new Qt3DRender::QParameter( "BB_SIZE", QSizeF( 100, 100 ), this ) )
+  , mViewportSize( new Qt3DRender::QParameter( "WIN_SCALE", QSizeF( 800, 600 ), this ) )
+{
+  addParameter( mSize );
+  addParameter( mViewportSize );
+
+  // Initialize with empty parameter.
+  mTexture2D = new Qt3DRender::QParameter( "tex0", QVariant(), this );
+  addParameter( mTexture2D );
+
+  // Blending for handling transparency
+  Qt3DRender::QBlendEquationArguments *blendState = new Qt3DRender::QBlendEquationArguments;
+  blendState->setSourceRgb( Qt3DRender::QBlendEquationArguments::SourceAlpha );
+  blendState->setDestinationRgb( Qt3DRender::QBlendEquationArguments::OneMinusSourceAlpha );
+
+  Qt3DRender::QBlendEquation *blendEquation = new Qt3DRender::QBlendEquation;
+  blendEquation->setBlendFunction( Qt3DRender::QBlendEquation::Add );
+
+  // Shader program
+  Qt3DRender::QShaderProgram *shaderProgram = new Qt3DRender::QShaderProgram( this );
+  shaderProgram->setVertexShaderCode( Qt3DRender::QShaderProgram::loadSource( QUrl( QStringLiteral( "qrc:/shaders/billboards.vert" ) ) ) );
+  shaderProgram->setFragmentShaderCode( Qt3DRender::QShaderProgram::loadSource( QUrl( QStringLiteral( "qrc:/shaders/billboards.frag" ) ) ) );
+  shaderProgram->setGeometryShaderCode( Qt3DRender::QShaderProgram::loadSource( QUrl( QStringLiteral( "qrc:/shaders/billboards.geom" ) ) ) );
+
+  // Render Pass
+  Qt3DRender::QRenderPass *renderPass = new Qt3DRender::QRenderPass( this );
+  renderPass->setShaderProgram( shaderProgram );
+  renderPass->addRenderState( blendState );
+  renderPass->addRenderState( blendEquation );
+
+  // without this filter the default forward renderer would not render this
+  Qt3DRender::QFilterKey *filterKey = new Qt3DRender::QFilterKey;
+  filterKey->setName( QStringLiteral( "renderingStyle" ) );
+  filterKey->setValue( "forward" );
+
+  // Technique
+  Qt3DRender::QTechnique *technique = new Qt3DRender::QTechnique;
+  technique->addRenderPass( renderPass );
+  technique->addFilterKey( filterKey );
+  technique->graphicsApiFilter()->setApi( Qt3DRender::QGraphicsApiFilter::OpenGL );
+  technique->graphicsApiFilter()->setProfile( Qt3DRender::QGraphicsApiFilter::CoreProfile );
+  technique->graphicsApiFilter()->setMajorVersion( 3 );
+  technique->graphicsApiFilter()->setMinorVersion( 1 );
+
+  // Effect
+  Qt3DRender::QEffect *effect = new Qt3DRender::QEffect( this );
+  effect->addTechnique( technique );
+
+  setEffect( effect );
+}
+
+void QgsPoint3DBillboardMaterial::setSize( const QSizeF size )
+{
+  mSize->setValue( size );
+}
+
+QSizeF QgsPoint3DBillboardMaterial::size() const
+{
+  return mSize->value().value<QSizeF>();
+}
+
+void QgsPoint3DBillboardMaterial::setViewportSize( const QSizeF size )
+{
+  mViewportSize->setValue( size );
+}
+
+QSizeF QgsPoint3DBillboardMaterial::windowSize() const
+{
+  return mViewportSize->value().value<QSizeF>();
+}
+
+void QgsPoint3DBillboardMaterial::setTexture2D( Qt3DRender::QTexture2D *texture2D )
```

Comment:
we can make this method private - I guess client code should use the higher level calls (setTexture2DFrom...())
or actually its body could be moved to setTexture2DFromTextureImage() as that's the only place where it's used and it's single line only...

Reply:
Moved to setTexture2DFromTextureImage.